src/responder/ifp/ifp_users.c

```diff
@@ -1029,6 +1031,339 @@ ifp_users_list_by_domain_and_name_recv(TALLOC_CTX *mem_ctx,
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Could not set up child handlers [%d]: %s\n",
+                  ret, sss_strerror(ret));
+            ret = ERR_P11_CHILD;
+            goto done;
+        }
+
+        /* Set up timeout handler */
+        tv = tevent_timeval_current_ofs(state->timeout, 0);
+        state->timeout_handler = tevent_add_timer(state->ev, req, tv,
+                                                  p11_child_timeout,
+                                                  req);
+        if (state->timeout_handler == NULL) {
+            ret = ERR_P11_CHILD;
+            goto done;
+        }
+        /* Now either wait for the timeout to fire or the child to finish */
+    } else { /* error */
+        ret = errno;
+        DEBUG(SSSDBG_CRIT_FAILURE, "fork failed [%d][%s].\n",
+              ret, sss_strerror(ret));
+        goto done;
+    }
+
+    return EAGAIN;
+
+done:
+    if (ret != EOK) {
+        PIPE_CLOSE(pipefd_from_child);
+        PIPE_CLOSE(pipefd_to_child);
+    }
+
+    return ret;
+}
+
+static struct tevent_req *
+ifp_users_find_by_valid_cert_done(int child_status,
+                                  struct tevent_signal *sige,
+                                  void *pvt)
+{
+    struct tevent_req *subreq;
+    struct tevent_req *req = talloc_get_type(pvt, struct tevent_req);
+    struct ifp_users_find_by_valid_cert_state *state = tevent_req_data(req,
+                                                  struct ifp_users_find_by_valid_cert_state);
+    bool valid = false;
+    errno_t ret;
+
+    PIPE_FD_CLOSE(state->io->read_from_child_fd);
+    PIPE_FD_CLOSE(state->io->write_to_child_fd);
+
+    if (WIFEXITED(child_status)) {
+        if (WEXITSTATUS(child_status) != 0) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  P11_CHILD_PATH " failed with status [%d]\n", child_status);
+        } else {
+            valid = true;
+        }
+    }
+
+    if (WIFSIGNALED(child_status)) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              P11_CHILD_PATH " was terminated by signal [%d]\n",
+              WTERMSIG(child_status));
+    }
+
+    if (valid) {
+        DEBUG(SSSDBG_TRACE_LIBS, "Certificate [%s] is valid.\n",
+                                  state->extra_args[0]);
+
+        subreq = cache_req_user_by_cert_send(state, state->ifp_ctx->rctx->ev,
+                                             state->ifp_ctx->rctx,
+                                             state->ifp_ctx->rctx->ncache, 0,
+                                             CACHE_REQ_ANY_DOM, NULL, state->derb64);
+        if (subreq == NULL) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "Unable to create subrequest!\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        tevent_req_set_callback(subreq, ifp_users_find_by_valid_cert_step, req);
+    } else {
+        DEBUG(SSSDBG_MINOR_FAILURE, "Certificate [%s] is not valid.\n",
+                                    state->extra_args[0]);
```

Comment:
Can we distinguish between invalid certification and error during validation process?